### PR TITLE
Allow access to all 12 TCP header flags, Decode 9th flag TH_AE (AccECN) as "a" stateless.

### PR DESCRIPTION
**Pull request by rscheff**
Provide the means to decode all 12 flag bits of the TCP header. 

While the Nonce Sum ("NS" in wireshark) was made historic (https://datatracker.ietf.org/doc/html/rfc3540 , https://datatracker.ietf.org/doc/status-change-ecn-signaling-with-nonces-to-historic/03/ ), that bit will become Accurate ECN ("AE") shortly (https://datatracker.ietf.org/doc/draft-ietf-tcpm-accurate-ecn/), which is an integral signaling protocol for the L4S framework.

Other utilities (packetdrill) have already been extended to allow the scripting with the "A" bit (or the full ACE 3-bit field as a shorthand; as state would be required to decide if to show the bits or the field, keeping this simple and only show individual fields - which makes a tcpdump a valid input for packetdrill still).

Also, there are two TCP option numbers for experimental use (253 and 254).

### CODE REVIEW

**Comment by rscheff** [comment]
Note that the tests fail, as all the test pcap file contain non-zero bits as reserved bits in the TCP headers.

I'll try to fix these binaries to follow the RFCs (initialize reserved bits with zero).

**Comment by infrastation** [comment]
Since a packet analyzer's purpose is to display what is on the wire (even and especially if what's on the wire is not correct), it might be better to leave the captures as they are and update the output instead.

**Comment by rscheff** [comment]
> Since a packet analyzer's purpose is to display what is on the wire (even and especially if what's on the wire is not correct), it might be better to leave the captures as they are and update the output instead.

Yes - I decided to do it this way, as editing the binaries was becoming painful. Reverted the one pcap, and adjusted the correct output instead now.

**Comment by rscheff** [comment]
Note that the output of tcpdump with this patch is equivalent with the fork of tcpdump in MacOS, being used for dev work on AccECN there, as well as the fork used by Google on Linux on their dev branches.

**Comment by rscheff** [comment]
> Since a packet analyzer's purpose is to display what is on the wire (even and especially if what's on the wire is not correct), it might be better to leave the captures as they are and update the output instead.

Question: The RFCs are relatively clear, that reserved fields should be initialized with zero until they get an assigned purpose. In the past there have been implementations, which would react too restrictive, when encountering non-zero reserved bits.

Should this be shown in the text output of tcpdump (rather than only when you fully decode and check all the packet header bits one-by-one?

**Comment by infrastation** [comment]
Quite often the most useful approach is to name the reserved bits using the `MBZ` prefix (as is done in `print-aos.c`, `print-radius.c` and `print-udld.c`) and/or to check the bit field using a "known allocated" bitmask and to print "bogus" if any unallocated bits are set (see `OFPPS_U` in `print-openflow-1.0.c`).

**Comment by rscheff** [comment]
Thank you. Is that a request to further improve this pull request? Or for a subsequent one?

Strictly speaking, these reserved bits are *NOT* MBZ (must be zero), but rather SBZ (RFC2119 SHOULD be zero) ;)

As all the other tcp header Flags are assigned a single Char for text decoding purposes, and the remaining 3 bits are undefined/reserved (a receiver or middlebox should not behave differently regardless of their state), I think this should not be part of this pull request.

As for the procedural stuff - is there any additional formal discussing needed for such a patch to go in? The Accurate ECN draft will be ready for WGLC very soon, but functional implementations of this protocol already exist (eg.  
https://developer.apple.com/videos/play/wwdc2022/10078/ ).

**Comment by infrastation** [comment]
The next step is that someone needs to study the changes with attention and to tell if everything looks ready [enough]. Eventually the WIP commits will need to be squashed together and merged. Please have some patience.

**Comment by infrastation** [comment]
For reference, respective IANA registry has moved [here](https://www.iana.org/assignments/tcp-parameters/tcp-parameters.xhtml#tcp-header-flags).

**Comment by infrastation** [comment]
Rebased on the current master branch, squashed, removed modifications to pre-existing packet captures and updated the tests to pass.

**Comment by infrastation** [comment]
Thank you for waiting. Please find below some more comments regarding the proposed code changes.
* The only change to `TH_ECNECHO` and `TH_CWR` is replacing a tab with a space, which is not consitent with the lines above and should not be done in this commit. Likewise, `tests/TESTLIST` should not lose the trailing newline.
* This change makes the TCP decoder recognize 9 bits (8 existing ones plus the new `AE` bit from the I-D above), not 12 flags; this way, currently the commit message does not match the changes and does not tell the rationale for the change, please improve that.
* The new `TH_RES1`, `TH_RES2` and `TH_RES3` constants are not used in any way; if these are an intended part of this change, please add respective code as discussed above, otherwise it will be right not to add the dead code.
* Most importantly, as far as the registry goes, the `AE` bit is neither allocated nor reserved. Surely there is the traditional chicken and egg problem here, but at the very least the comment for the new constant should tell the I-D name, revision and expiry date. As a matter of fact, a significant part of this source code maintenance comes from early implementations of I-Ds that either changed the protocol in later revisions or lapsed, but the code was left in place. Please address that in the best way you can.

**Comment by rscheff** [comment]
> Thank you for waiting. Please find below some more comments regarding the proposed code changes.

Sorry - apparently this response never reached me.

> * The only change to `TH_ECNECHO` and `TH_CWR` is replacing a tab with a space, which is not consitent with the lines above and should not be done in this commit. Likewise, `tests/TESTLIST` should not lose the trailing newline.

The other way around, making this consistent with the other lines. Since this is a style change and unrelated, i reverted this and leave the differently styled lines as they are. A trailing newline should now be part of tests/TESTLIST.

> * This change makes the TCP decoder recognize 9 bits (8 existing ones plus the new `AE` bit from the I-D above), not 12 flags; this way, currently the commit message does not match the changes and does not tell the rationale for the change, please improve that.

Fixed the naming of this pull request

> * The new `TH_RES1`, `TH_RES2` and `TH_RES3` constants are not used in any way; if these are an intended part of this change, please add respective code as discussed above, otherwise it will be right not to add the dead code.

Removed.

> * Most importantly, as far as the registry goes, the `AE` bit is neither allocated nor reserved. Surely there is the traditional chicken and egg problem here, but at the very least the comment for the new constant should tell the I-D name, revision and expiry date. As a matter of fact, a significant part of this source code maintenance comes from early implementations of I-Ds that either changed the protocol in later revisions or lapsed, but the code was left in place. Please address that in the best way you can.

I added relevant referecences to the architeture RFCs under which AccECN is being standardized (currently WGLC; RFC by the end of the year likely) into the comment in tcp.h.

**Comment by rscheff** [comment]
When looking at the CI failures, it seems they fail due to some connectivity issues, but not the changes in this pull request...

**Comment by infrastation** [comment]
Looks fine to me with the added commit. I am going to merge this pull request before the end of this Sunday, if anybody has any other remarks, please make these soon. If anybody merges this pull request before me, please do not forget to squash and to make sure the commit message looks good.

**Comment by infrastation** [comment]
@rscheff, please see issue #1069 about the man page. Would it be better to update the man page in the same commit?

**Comment by fxlb** [comment]
The man page should be updated to add the new flag.

**Comment by fxlb** [comment]
It may be useful to point out that the new 'A' flag is not ACK (ACK is '.').

**Comment by rscheff** [comment]
> I checked in a minor update to the man page. However, reading through this, I would like to understand where the parser code lives that interprets "tcp[tcpflags] & tcp-cwr != 0". The tcpflags field is really 12 bit wide (octets 12&13), and to align with what is currently available, tcp-ae should also be a named flag (0x100) to match against this 12 bit field...

Found that the parser actually lives in libpcap; allowing a token like "tcp[tcpflags]" to extract the full tcp flags field would have to go there (and be followed by corresponding man page changes). This however should not block this pull request. Once the new parser/grammer to allow that (special casing) is ready, I'll make a pull request for libpcap.

**Comment by infrastation** [comment]
The man page requires more work before merging. Your branch now has the fixup again. Let's not touch libpcap until the registry has settled, it is a pain to change the filter syntax after it has been published.

File: tcpdump.1.in
Comment (fxlb) [inline tcpdump.1.in:1445]: Two `|A|` are confusing.

File: tcpdump.1.in
Comment (fxlb) [inline tcpdump.1.in:1462]: Same here;

File: tcpdump.1.in
Comment (infrastation) [inline tcpdump.1.in:1445]: This packet diagram cannot convey the meaning using "A" for two different flags. Either it will have to use more characters per bit, or the flags field needs be unpacked only in a separate 12-bit diagram.

**Review by infrastation — CHANGES_REQUESTED** [review]

File: tcpdump.1.in
Comment (fxlb) [reply tcpdump.1.in:1445]: Something like?
flags

Comment (fxlb) [reply tcpdump.1.in:1445]: And ` W (CWR)` in the man page should be ` C (CWR)`.

Comment (fxlb) [reply tcpdump.1.in:1445]: > And ` W (CWR)` in the man page should be ` C (CWR)`.

No, bad idea. It's 'W' in the code.

Comment (infrastation) [reply tcpdump.1.in:1445]: Vertical labels could work, yes.

Comment (rscheff) [reply tcpdump.1.in:1445]: So - variant one, use the single-char appreviations here that tcpdump would show in the output:
|A|W|E|U|.|P|R|S|F

or expand to 3 lines, and in both cases have  a short table translating the single char to RFC abbreviations?

Comment (rscheff) [reply tcpdump.1.in:1445]:
```
 There are 9 bits in the control bits section of the TCP header:
 .I AE | CWR(W) | ECE | URG | ACK(.) | PSH | RST | SYN | FIN
 
|  HL   |rsvd |A|W|E|U|.|P|R|S|F|        window size            |
```

Comment (fxlb) [reply tcpdump.1.in]: > Vertical labels could work, yes.

Adapted from a wikipedia diagram.

Comment (infrastation) [reply tcpdump.1.in]: Use of `.` to indicate `ACK` is a peculiarity of the protocol analyzer software; the motivation behind that may or may not be lost in time, but in any case the discrepancy has fossilized and it would be wrong to fix it in the software. It would also be wrong to diverge from the normative specification's version of the packet diagrams in the man page. So the best balance seems to be making the man page closer to the spec and preserving the traditional remarks such as "The `.' means the ACK flag was set." and the likes.

Comment (fxlb) [reply tcpdump.1.in]: To avoid confusion 'A' with ACK, could the new flag for 'Accurate ECN' be 'N'?

Comment (infrastation) [reply tcpdump.1.in]: Do you mean in the `[SF.PUE]` output syntax? It is not difficult to imagine how "A" could easily confuse less experienced users. Bit 7 used to mean NS in the past, and N is the only character that appears in both "Accurate ECN" and "NS" (not in "AE" though). Or perhaps it could be just "7", which would also simplify representation of the remaining three bits. Perhaps it would be best to bring this up on the mailing list.

But the man page should be consistent with the spec, and the I-D says:
```
     +=====+==============+===========+==============================+
     | Bit | Name         | Reference | Assignment Notes             |
     +=====+==============+===========+==============================+
     | 7   | AE (Accurate | RFC XXXX  | Previously used as NS (Nonce |
     |     | ECN)         |           | Sum) by [RFC3540], which is  |
     |     |              |           | now historic [RFC8311]       |
     +-----+--------------+-----------+------------------------------+
```

Comment (fxlb) [reply tcpdump.1.in]: > Do you mean in the `[SF.PUE]` output syntax?

Yes, 'N' like 'Accurate EC**N**'. The 'W' for C**W**R and the '.' for ACK show it is not always the first character of the flag name.

Comment (infrastation) [reply tcpdump.1.in]: It does not work in every case, but it is nice when it works. For the avoidance of doubt, if anyone is going to suggest "Æ" for the single-character syntax, that would not work well for all users.

Comment (fxlb) [reply tcpdump.1.in]: Yes, keep ASCII output only.

Comment (rscheff) [reply tcpdump.1.in]: > To avoid confusion 'A' with ACK, could the new flag for 'Accurate ECN' be 'N'?

Other testing/tooling code (wireshark, packetdrill) is already using the "A" character here (and 0..7 when the use of the ACE field was detected in the 3WHS). Since the use of "." for ACK, as spearheaded by tcpdump, is a well recognized standard, people misinterpreting "A" here will certainly learn this fast - and remember it better for that mistake.

And I thought about Unicode, but since tcpdump is supposed to work on a high variety of platforms and terminals, introducing i8n/mapping just here is not ideal.

Comment (fxlb) [reply tcpdump.1.in]: Where Wireshark use "A"?  tshark gives:

```
...1 .... .... = Accurate ECN: Set
```

Comment (rscheff) [reply tcpdump.1.in]: > Do you mean in the `[SF.PUE]` output syntax? It is not difficult to imagine how "A" could easily confuse less experienced users. Bit 7 used to mean NS in the past, and N is the only character that appears in both "Accurate ECN" and "NS" (not in "AE" though). Or perhaps it could be just "7", which would also simplify representation of the remaining three bits. Perhaps it would be best to bring this up on the mailing list.
> 
> But the man page should be consistent with the spec, and the I-D says:
> 
> ```
>      +=====+==============+===========+==============================+
>      | Bit | Name         | Reference | Assignment Notes             |
>      +=====+==============+===========+==============================+
>      | 7   | AE (Accurate | RFC XXXX  | Previously used as NS (Nonce |
>      |     | ECN)         |           | Sum) by [RFC3540], which is  |
>      |     |              |           | now historic [RFC8311]       |
>      +-----+--------------+-----------+------------------------------+
> ```

As just mentioned, using numerals here would conflict with existing tooling/testing code, where 0..7 represents the different codepoints that the ACE field can have after the AccECN handshake. Also, this bit is canonically the 9th bit in the 12-bit wide field reserved for TCP header flags. This discussion would have been good to have last year, when the set of patches got upstreamed across a number of tools.

In any case, I updated the man page to show the full RFC abbreviations, and adjusted the indentation of the line describing the bits and their single character representation.

Comment (rscheff) [reply tcpdump.1.in]: > Where Wireshark use "A"? tshark gives:
> 
> ```
> ...1 .... .... = Accurate ECN: Set
> ```

See patch https://github.com/wireshark/wireshark/commit/ecefcf880121961c03d2ec05ad378c82e951e0a2

The ASCII characters used in the abbreviated decoding are "RRRACEUAPRSF" - so both ACK and AE use "A"; wireshark will however decode the full ACE field typically into one of the numericals 0..7 during the session, or use the "long" abbreviation:

    4   2.055806 192.168.20.233 → 192.168.20.216 TCP  60 0 00:0c:29:f7:80:12 08:00:27:26:77:7f  29806 → 22 [SYN, ECE, CWR, AE] Seq=0 Win=65535 Len=0 MSS=1460 WS=64 SACK_PERM TSval=155081733 TSecr=0

Comment (fxlb) [reply tcpdump.1.in]: Having 2 "A" when printing someting like "R--AC-UA--S-" where each "A" is at a specific place is different.
It is not the way tcpdump works.

Comment (rscheff) [reply tcpdump.1.in]: Yes. tcpdump (and packetdrill) have been using "." for ACK and packetdrill is using "A" for AE, and additionally 0..7 for the full ACE field (comprised of AE, CWR, ECE).

Comment (infrastation) [reply tcpdump.1.in]: Thank you for preparing the update, the formatting looks consistent. In order to avoid another round of discussion later, let me suggest to bring the naming/labelling aspect on the mailing list before merging these changes.

Comment (guyharris) [reply tcpdump.1.in]: > Use of . to indicate ACK is a peculiarity of the protocol analyzer software; the motivation behind that may or may not be lost in time,

If we look at f555c163f90c9de17ebcef8313f86404c5174ca9, which was the commit that created the tcpdump repository (from back when it used CVS, before being converted to Git), the print-tcp.c in that commit (which is the same as the tcpdump 3.4 print-tcp.c) has:

```
	if ((flags = tp->th_flags) & (TH_SYN|TH_FIN|TH_RST|TH_PUSH)) {
		if (flags & TH_SYN)
			putchar('S');
		if (flags & TH_FIN)
			putchar('F');
		if (flags & TH_RST)
			putchar('R');
		if (flags & TH_PUSH)
			putchar('P');
	} else
		putchar('.');
```

as the code that dissects the flags field.  ".", at that point, meant "this is a boring data transfer or ACK-only packet", as opposed to the initial SYN, the SYN+ACK response, an RST, *or* a "here's a PSH" packet.  (URG packets are indicated by "urg <urgent pointer value>" in the output.)

Presumably "." was chosen because it doesn't stand out much.

The commit that switched to using a table of flag values/characters and the "flag bits to string" routine was commit 9427d40c14077527b18af1bedeae39457b2981fe:

```
Author: Hannes Gredler <hannes@gredler.at>
Date:   Mon Jan 29 09:59:42 2007 +0000

    change the TCP printer to print new-style order:
      - seperate fields by comma
      - use [] for flags and options sets
      - print a trailing length field

    make use of tok2str() and bitttok2str()
    move port definitions into tcp.h
```

Presumably "." was chosen so that boring data transfers and ACK-only packets will still show up as ".".

One could argue that explicitly showing the ACK flag means that, for example, a boring data transfer that *doesn't* acknowledge anything - which, as I remember, should not happen - can be distinguished from one that does.

Comment (rscheff) [reply tcpdump.1.in]: I would argue the use of „.“ for the ACK flag is well established in the trade, so that a sporadic „A“ would not indroduce much ambiguity - probably more curiosity to learn what this AccECN is all about, and the larger L4S project with TCP Prague and DualQ AQM ;)

**Comment by rscheff** [comment]
Any update on taking this on? Where are we here?

**Comment by infrastation** [comment]
Please raise this on the mailing list, as suggested. Even if there is no feedback, it will be useful to know.

**Comment by rscheff** [comment]
I’ve tried to subscribe to the mailing list given the instructions https://www.tcpdump.org/#lists, but my post doesn’t show up here https://seclists.org/tcpdump/2023/q3/

Nor was there any reply from majordomo (and derivatives) which you’d normally get when subscribing…

The traffic volume also appears to be so low, that I question how active this is (or if the majordomo instance may be broken for a few months ☹


Richard Scheffenegger
Consulting Solution Architect
NAS & Networking

NetApp
+43 1 3676 811 3157 Direct Phone
+43 664 8866 1857 Mobile Phone
***@***.******@***.***>

**Comment by infrastation** [comment]
The recent migration to Mailman 3 had issues, some of which still stand. I have received your message, which means other subscribers have received it too. Thank you for posting, I propose to merge the changes if nobody objects by the end of August.

**Comment by fxlb** [comment]
> I propose to merge the changes if nobody objects by the end of August.

I think this is premature, as the mailing list isn't working properly. I haven't received the reply I sent.
Nor does it appear on https://seclists.org/tcpdump/2023/q3/.
If interested parties can't reply, we can't conclude on the choice of the letter representing the new flag.

**Comment by rscheff** [comment]
This is still in limbo, without me understanding if there is any constructive path forward here.

In the meantime, other areas where the TCP header is decoded suggested to use "a" for AE while there "A" is for ACK...

Updated this patch accordingly...

**Comment by tuexen** [comment]
> This is still in limbo, without me understanding if there is any constructive path forward here.
> 
> In the meantime, other areas where the TCP header is decoded suggested to use "a" for AE while there "A" is for ACK...
> 
> Updated this patch accordingly...

Isn't an ACK displaying as `.` in tcpdump? How can I see it printed as an 'A'?
I would really like to see all twelve flags being displayed if they are non zero. In addition to the AE flags, these would be three flags which are currently being reserved. But that does not mean that they don't show up non-zero on the network.

**Comment by infrastation** [comment]
@rscheff, thank you for waiting. This matter is taking quite a while longer than it should.

The points you made earlier have been in the back of my mind for several months, but just as I was ready to respond to your latest post to tcpdump-workers, the mailing list went down. As it seems to require more time to get fixed, let's try to use this pull request comments again for now.

First of all, I agree that the use of lowercase letters is a good idea. But let me suggest to avoid even "a" for AccECN to eliminate as much confusion with ACK as possible. One way to achieve this could be as follows:
```
F -- FIN
S -- SYN
R -- RST
P -- PSH
. -- ACK
U -- URG
E -- ECE (ECN-Echo)
W -- CWR

p -- future development based on PSH, if any
u -- future development based on URG, if any
e -- AE (AccECN)
w -- future development based on CWR, if any
```
Do you agree this would work a bit better long-term? I can update and rebase the changes accordingly if you want.

**Comment by rscheff** [comment]
Yes, I also reconsidered - while commonality between different tools was in the back of my mind, I don't think this should stop making AccECN decoding available in as many tools as possible. In the meantime, another tool got AccECN support, also with its own nomenclature (https://github.com/Netflix/read_bbrlog/pull/4/commits/b9f66a4bb7773cbe070753177b478f6705d3b992 if anyone is interested) - as long as we decode it, I'm fine.

The tests (including old test files with randomized header data) would need adjustement too. If you can take care of this, it would be great. In the meantime, we found that certain NIC drivers on the send path need patches too, for efficient TSO support of AccECN.

**Comment by fxlb** [comment]
We could choose:
1) The Denis proposal "***e***"
Or to keep uppercases:
2) "More Accurate Explicit ***C***ongestion Notification (E***C***N)", thus "***C***".
3) "More Accurate Explicit Congestion ***N***otification (EC***N***)", thus "***N***".

**Comment by rscheff** [comment]
C conflicts with CWR, which has been choosen in some tooling; 
N would retain the (now historic) same abbreviation as the former Nonce Sum... (RFC3540). 

Finally, if AccECN does see more widespread adoption to make it worthwhile to spend more effort on this, a stateful decoding into the ACE counter (0..7), like what packetdrill does, could be added at a later point.

In short, I'm agnostic to "N" or "e", "C" would get me confused with "CWR" as thats the mapping my other tools use.

**Comment by fxlb** [comment]
Ok to remove the confusing "C".

Remain "N" or "e",

**Comment by fxlb** [comment]
Any other opinions?

**Comment by fxlb** [comment]
Or ***M*** like ***M***ore Accurate Explicit Congestion Notification (ECN)?

**Comment by infrastation** [comment]
Both ECN-Echo and AccECN have E and C and N, so it would be far from obvious which one is "E" and which one is "N". But "accurate" is a synonym of "fine" and "delicate", and "e" is the closest to "fine E" in ASCII. That's the way I can remember it.

**Comment by fxlb** [comment]
Ok, "N" seems confusing.

We can choose "e" if there are no objections.

**Comment by infrastation** [comment]
Changed the "a" to "e", rebased on the master branch and updated all tests related to these changes. This revision is expected to pass all checks.

**Comment by fxlb** [comment]
Why `tests/bgp_mp_reach_nlri-oobr.pcap` was updated?

**Comment by fxlb** [comment]
I think it should not be updated, but the output must be updated from `Flags [none]` to `Flags [e]`.

**Comment by fxlb** [comment]
Fixed.

**Comment by fxlb** [comment]
> Note that the output of tcpdump with this patch is equivalent with [...] the fork used by Google on Linux on their dev branches.

@rscheff What is the link of this tcpdump fork?

**Review by infrastation — APPROVED** [review]

**Comment by infrastation** [comment]
Should be squashed (and perhaps given a better commit message) to merge. The expansion of TCP flags into 2/3 characters in the TCP header diagrams is not a necessity, but looks useful anyway.

**Comment by infrastation** [comment]
Squashed, fixed the change log entry and the commit message. This is going to be merged as soon as the CI passes.

**Comment by infrastation** [comment]
Thank you for waiting. The updated man page is now [live](https://www.tcpdump.org/manpages/tcpdump.1.html).

**Pull request merged**
